incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -372,36 +376,49 @@ func (r *ConfigReconciler) validateSingletonName(ctx context.Context, nm string)
 	inst.Status.Conditions = nil
 	inst.Status.Conditions = append(inst.Status.Conditions, condition)
 
-	if err := r.writeSingleton(ctx, inst); err != nil {
-		return err
-	}
-
-	return fmt.Errorf("Error while validating singleton name: %s", msg)
+	return false, r.writeSingleton(ctx, inst)
 }
 
-// addTypeStatus adds Status.Types for types configured in the spec. Only the status of
+// setTypeStatuses adds Status.Types for types configured in the spec. Only the status of
 // types in `propagate` and `remove` modes will be recorded. The Status.Types
-// is sorted in alphabetical order based on APIVersion.
-func (r *ConfigReconciler) addTypeStatus(inst *api.HNCConfiguration) {
+// is sorted in alphabetical order based on APIVersion and Kind.
+func (r *ConfigReconciler) setTypeStatuses(inst *api.HNCConfiguration) {
+	statuses := []api.TypeSynchronizationStatus{}
 	for _, ts := range r.Forest.GetTypeSyncers() {
-		if r.activeGVKs[ts.GetGVK()] && ts.GetMode() != api.Ignore {
-			r.addNumPropagatedObjects(ts.GetGVK(), ts.GetNumPropagatedObjects(), inst)
+		// Don't output a status for any reconciler that isn't explicitly listed in the spec
+		gvk := ts.GetGVK()
+		if !r.activeGVKs[ts.GetGVK()] {
+			continue
 		}
+
+		// Initialize status
+		apiVersion, kind := gvk.ToAPIVersionAndKind()
+		status := api.TypeSynchronizationStatus{
+			APIVersion: apiVersion,
+			Kind:       kind,
+			Mode:       ts.GetMode(), // may be different from the spec if it's implicit
```

Comment:
Question: How is this GetMode() up to date?

Reply:
On line 329, we validate the mode (e.g. change it from the empty string to `propagate`) for a new controller. When we update an existing controller, it happens on line 250. Does that make sense?

Reply:
Thank you for the explanation! Ah I see, this is the forest version and it may not be up to date (because of the lag in between the mode change) but it will eventually match the spec. It always reflects the mode Object controller is using.

Reply:
That's exactly it. This was the cause of the flakiness in the tests - we set the spec to "ignore," but then _before_ that could take effect, we ran a test that assumed the object controller was already ignoring stuff. But it wasn't.

Reply:
Got you, yes, that's why you added `typeHasStatus()` before checking the object propagation.

Reply:
Yup!